Plutarch/FFI.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.FFI (
+  foreignExport,
+  foreignImport,
+  unsafeForeignExport,
+  unsafeForeignImport,
+) where
+
+import Data.Kind (Constraint, Type)
+import Data.Text (Text)
+import GHC.TypeLits (TypeError)
+import qualified GHC.TypeLits as TypeLits
+import qualified Generics.SOP as SOP
+import Plutarch.Bool (PBool)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  ClosedTerm,
+  PDelayed,
+  PType,
+  RawTerm (RCompiled),
+  Term (Term),
+  TermResult (TermResult),
+  asClosedRawTerm,
+  compile',
+  (:-->),
+ )
+import Plutarch.Internal.PlutusType (PlutusType (PInner))
+import Plutarch.String (PString)
+import Plutus.V1.Ledger.Scripts (Script (unScript), fromCompiledCode)
+import PlutusTx.Code (CompiledCode, CompiledCodeIn (DeserializedCode))
+import PlutusTx.Prelude (BuiltinString)
+import UntypedPlutusCore (fakeNameDeBruijn)
+import qualified UntypedPlutusCore as UPLC
+
+data ForallPhantom :: Type
+data PhorallPhantom :: PType
+
+foreignExport :: PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom => ClosedTerm p -> CompiledCode t
+foreignExport = unsafeForeignExport
+
+foreignImport :: PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom => CompiledCode t -> ClosedTerm p
+foreignImport = unsafeForeignImport
+
+unsafeForeignExport :: ClosedTerm p -> CompiledCode t
```

Comment:
I think we can possibly use `Any` instead of `t` here to make it safe.

Reply:
If you're using this from PlutusTx code you'd use `unsafeCoerce` to turn it into the real thing.

Reply:
A typical `PlutusTx` user has no clue about data type representation of various types, and even less clue how they match between PlutusTx and Plutarch. Even I was surprised to find the `Bool` representations were mismatched. What's more, a mismatch would not be caught at compile time and not necessarily even at the first run-time: it may be just lurking in a rarely taken execution path. After it's submitted on chain, in a validator script guarding millions of dollars. We can't leave a bomb like this in users' hands.

Reply:
Isn't this an argument against having `unsafeForeignExport`? The same arguments applies here as with `unsafeForeignImport`.

Reply:
`foreignExport :: ClosedTerm p -> CompiledCode Any` wouldn't be unsafe though.

Reply:
I think there's an unsafe function`CompiledCode a -> CompiledCode b`. You could always do `unsafeCoerce` inside a PlutusTx script.